contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,775 @@
+      retrieveNXMPayment(totalPremiumInNXM, params.commissionRatio, params.commissionDestination);
+    } else {
+      retrievePayment(
+        premiumInPaymentAsset,
+        params.paymentAsset,
+        params.commissionRatio,
+        params.commissionDestination
+      );
+    }
+
+    // push the newly created cover
+    _coverData.push(CoverData(
+        params.productId,
+        params.payoutAsset,
+        0 // amountPaidOut
+      ));
+
+    uint coverId = _coverData.length - 1;
+    ICoverNFT(coverNFT).safeMint(params.owner, coverId);
+
+    // Enable this when cover amount tracking is necessary
+    if (coverAmountTrackingEnabled) {
+      totalActiveCoverInAsset[params.payoutAsset] += params.amount;
+    }
+
+    emit CoverBought(coverId, params.productId, 0, msg.sender, params.ipfsData);
+    return coverId;
+  }
+
+  function _buyCover(
+    BuyCoverParams memory params,
+    uint coverId,
+    PoolAllocationRequest[] memory allocationRequests
+  ) internal returns (uint totalPremiumInNXM) {
+
+    // convert to NXM amount
+    uint nxmPriceInPayoutAsset = pool().getTokenPrice(params.payoutAsset);
+    uint remainderAmountInNXM = 0;
+    uint totalCoverAmountInNXM = 0;
+
+    uint _coverSegmentsCount = _coverSegments[coverId].length;
+
+    for (uint i = 0; i < allocationRequests.length; i++) {
+
+      uint requestedCoverAmountInNXM
+        = allocationRequests[i].coverAmountInAsset * 1e18 / nxmPriceInPayoutAsset + remainderAmountInNXM;
+
+      (uint coveredAmountInNXM, uint premiumInNXM, uint rewardsInNXM) = allocateCapacity(
+        params,
+        coverId,
+        stakingPool(allocationRequests[i].poolId),
+        requestedCoverAmountInNXM
+      );
+
+      // apply the global rewards ratio and the total Rewards in NXM
+      tokenController().mintStakingPoolNXMRewards(rewardsInNXM, allocationRequests[i].poolId);
+
+      remainderAmountInNXM = requestedCoverAmountInNXM - coveredAmountInNXM;
+      totalCoverAmountInNXM += coveredAmountInNXM;
+      totalPremiumInNXM += premiumInNXM;
+
+      coverSegmentAllocations[coverId][_coverSegmentsCount].push(
+        PoolAllocation(
+          allocationRequests[i].poolId,
+          SafeUintCast.toUint96(coveredAmountInNXM),
+          SafeUintCast.toUint96(premiumInNXM)
+        )
+      );
+    }
+
+    // priceRatio is normalized on a per year basis (eg. 1.5% per year)
+    uint16 priceRatio = SafeUintCast.toUint16(
+      divRound(
+        totalPremiumInNXM * PRICE_DENOMINATOR * MAX_COVER_PERIOD / params.period,
+        totalCoverAmountInNXM
+      )
+    );
+
+    _coverSegments[coverId].push(CoverSegment(
+        SafeUintCast.toUint96(totalCoverAmountInNXM * nxmPriceInPayoutAsset / 1e18), // amount
+        uint32(block.timestamp + 1), // start
+        SafeUintCast.toUint32(params.period), // period
+        priceRatio,
+        false, // expired,
+        globalRewardsRatio
+      ));
+
+    return totalPremiumInNXM;
+  }
+
+  function allocateCapacity(
```

Comment:
looks like this can now be inlined

Reply:
```CompilerError: Stack too deep, try removing local variables.
   --> contracts/modules/cover/Cover.sol:248:11:
    |
248 |           coverId,
    |           ^^^^^^^


Error HH600: Compilation failed

```

Attempted, stack too deep. imo not worth it now, the function gets a lot longer as well.